src/OrbitGl/LineGraphTrack.cpp

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "LineGraphTrack.h"
+
+#include <GteVector.h>
+
+#include <algorithm>
+#include <cstdint>
+#include <iterator>
+
+#include "Geometry.h"
+#include "TextRenderer.h"
+#include "TimeGraph.h"
+#include "TimeGraphLayout.h"
+
+namespace orbit_gl {
+
+template <size_t Dimension>
+float LineGraphTrack<Dimension>::GetLabelYFromValues(
+    const std::array<double, Dimension>& values) const {
+  float content_height = this->size_[1] - this->layout_->GetTrackTabHeight() -
+                         this->layout_->GetTrackBottomMargin() - this->GetLegendHeight();
+  float base_y = this->pos_[1] - this->size_[1] + this->layout_->GetTrackBottomMargin();
+  double min = this->GetGraphMinValue();
+  double inverse_value_range = this->GetGraphInverseValueRange();
+  std::array<float, Dimension> normalized_values;
+  std::transform(values.begin(), values.end(), normalized_values.begin(),
+                 [min, inverse_value_range](double value) {
+                   return static_cast<float>((value - min) * inverse_value_range);
+                 });
+
+  if (Dimension == 1) return base_y + normalized_values[0] * content_height;
+  return base_y + content_height / 2.0f;
+}
+
+template <size_t Dimension>
+void LineGraphTrack<Dimension>::DrawSeries(Batcher* batcher, uint64_t min_tick, uint64_t max_tick,
+                                           float z) {
+  auto entries_affected_range_result =
+      this->series_.GetEntriesAffectedByTimeRange(min_tick, max_tick);
+  if (!entries_affected_range_result.has_value()) return;
+
+  typename MultivariateTimeSeries<Dimension>::Range& entries{entries_affected_range_result.value()};
+
+  double min = this->GetGraphMinValue();
+  double inverse_value_range = this->GetGraphInverseValueRange();
+
+  auto current_iterator = entries.begin;
+  uint64_t current_time = current_iterator->first;
+  std::array<float, Dimension> current_normalized_values;
+  std::transform(current_iterator->second.begin(), current_iterator->second.end(),
+                 current_normalized_values.begin(), [min, inverse_value_range](double value) {
+                   return static_cast<float>((value - min) * inverse_value_range);
+                 });
+
+  while (current_iterator != entries.end) {
+    auto next_iterator = std::next(current_iterator);
```

Comment:
We will have a case of next_iterator = entries.end, but this is not a problem why?

Reply:
If we specify a time range [min_tick, max_tick], then `entries.end` points to the next time series entry of max_tick, or the last time series entry if the next does not exists.

Reply:
In general, I would expect end() in a datastructure to be a nullptr. This is confusing me. At least we should make a comment and a CHECK() for nullptr, what do you think?

Reply:
We had the following comments
```
  // If there is no overlap between time range [min_time, max_time] and [StartTimeInNs(),
  // EndTimeInNs()], return std::nullopt. Otherwise return a range of entries affected by the time
  // range [min_time, max_time] where:
  // * `Range::begin` points to the entry with the time key right before the time range (min_time,
  // max_time) if exists; otherwise points to the fist entry.
  // * `Range::end` points to the entry with the time key right after the time range (min_time,
  // max_time) if exists; otherwise points to the last entry.
```
behind the `MultivariateTimeSeries::GetEntriesAffectedByTimeRange` method which explains what the range refers to.

And the reason for taking "the next or last" element rather than `nullptr` as the range end is to simply the logic for drawing time series. 
Also `if (!entries_affected_range_result.has_value()) return;` already guarantees that the `range.end != nullptr` so the `CHECK` would be redundant.

Reply:
The CHECK was only about code legibility, I understood that couldn't be nullptr. If not, we should have crashed.

I'm still not sure about this, I'm used to always have end as a nullptr and in this case looping until next is equal to end, instead of current = end. @florian-kuebler what do you think?

Reply:
We used to return two iterators for the `GetEntriesAffectedByTimeRange` method, i.e., the first one points to the previous or first entry and the second one points to the next or last entry. 

The `Range` is proposed to serve as the return value of the `GetEntriesAffectedByTimeRange` method at that time.

Reply:
I think it is fine as it is. But I understand Pablo's concerns. Usually we have this `end()` iterator (its not a pointer though) to check if we are out of our data structure. Maybe we can simply rename the `begin`/`end` fields here?
What about `start_inclusive`/`end_exclusive`?

Reply:
In this case, it would be `start_inclusive` and `end_inclusive`. I'm fine with the naming. Thanks for the suggestion!